### PR TITLE
refactor: switch to DAG-CBOR

### LINKED ISSUE #85: Switch to dag-cbor
# Problem

This library is very old, and remembers the time before we had dag-cbor.
It uses  stringified JSON put in data field of `dag-pb` which is not only inefficiency and a technical debt, but an antipattern these days.

# Solution

- Remove use of `ipfs.object` API and JSON in `dag-pb`
  - Use `ipfs.block`  (`ipfs.dag` may change, `block` won't) and space-efficient [dag-cbor](https://ipld.io/docs/codecs/known/dag-cbor/) (https://www.npmjs.com/package/@ipld/dag-cbor)

### PR DESCRIPTION
**Pull request by lidel**
This PR aims to close #85 and switch b-tree away from expensive JSON stored in `Data` field of `dag-pb` and deprecated `object` RPC API. 

## TLDR: ~30% DAG size reduction

DAG-PB was 399424 bytes (~400MiB)
DAG-CBOR is 281089143  bytes (~281MiB) 

:point_right: this means ~30% reduction :ok_hand: 

## TODO

- [x] switch from `ipfs.object.get` to `ipfs.block.get`
- [x] switch from expensive `ipfs.object.put`  to CAR writer
  - [x] create a CAR with entire b-tree
  - [x] mark final root node as the CAR root
- [x] remove use of DAG-PB and JSON serialized as string
- [x] create DAG-CBOR nodes
  - [x] use CID CBOR tag in `links` so b-tree can be traversed by generic IPLD libraries and `ipfs dag stat` in Kubo
  - [x] remove string fields that indicate leaf nodes, use presence of `data` and `links` to reason about node type (storage savings) 
- [x] bump aegir and switch to ESM
- [x] switch from legacy JS libs to modern https://github.com/multiformats/js-multiformats/
- [x] regenerate dataset and confirm `node example/lookup.js 9.9.9.9` returns same result
- [x] detect when `ipfs` is not CoreAPI instance, but a string, and if it is a valid gateway URL, use it for fetching raw blocks as [application/vnd.ipld.raw](https://www.iana.org/assignments/media-types/application/vnd.ipld.raw)
- [ ] update source dataset and regenerate b-tree, so users not only get storage savings thanks to DAG-CBOR, but also get updated GeoIP database
- [ ] cleanup add retry  if gateway timeouts
- [ ] solve issue with `ip` dependency requiring native node module

## Tests

- [x] fix / update  `npm test`
-[ ]  TBD - maybe we should only support gateways?
  - [ ] confirm it works with `ipfs.block.get` from Kubo
  - [ ] confirm it works with `ipfs.block.get` from JS-IPFS
- [ ] confirm it works w
  - [ ] https://github.com/ipfs/public-gateway-checker (refactor to use gateway)
  - [ ] https://github.com/ipfs/ipfs-webui (refactor to use `availableGatewayUrl` and pass it similar to how we do it on  [explore page](https://github.com/ipfs/ipfs-webui/blob/f3a886568a9045318ee340c1b5ebe811f7d9de04/src/explore/ExploreContainer.js))

### CODE REVIEW

**Comment by lidel** [comment]
I'm merging this to unblock @whizzzkid  to clean up remaining items before we can release:
1. remove dependency on  `os` by doing the build step
3. fix npm test in [retry PR](https://github.com/ipfs-shipyard/ipfs-geoip/pull/95)

When we have that, I'll regenerate dataset and make the release.
